### ISSUE TITLE
Query mapper bug(s)?

### ISSUE THREAD

**Issue by n9** [linq2db/linq2db#873]
Let's have a sample query:

```C#
var q = Db.Context.GetTable<Product>();
Db.Context.GetTable<SuperProduct>()
	.Select(e => new
	{
		Fields = new
		{
			UiLabel = " " +
				e.NameEnUs,
			ProductCount = q.Where(_ => _.SuperProduct == e).Count()
		},
	})
	.Where(_ => _.Fields.UiLabel.Contains("a")).Dump();
```
This query fails with: 
> variable 'e' of type 'Entities.Orm.SuperProduct' referenced from scope '', but it is not defined
```
at System.Linq.Expressions.Compiler.VariableBinder.Reference(ParameterExpression node, VariableStorageKind storage)
at System.Linq.Expressions.Compiler.VariableBinder.VisitParameter(ParameterExpression node)
at System.Linq.Expressions.ParameterExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.VisitBinary(BinaryExpression node)
at System.Linq.Expressions.BinaryExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.Compiler.VariableBinder.VisitUnary(UnaryExpression node)
at System.Linq.Expressions.UnaryExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.VisitArguments(IArgumentProvider nodes)
at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.VisitArguments(IArgumentProvider nodes)
at System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
at System.Linq.Expressions.ExpressionVisitor.VisitNew(NewExpression node)
at System.Linq.Expressions.NewExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
at System.Linq.Expressions.ExpressionVisitor.VisitNew(NewExpression node)
at System.Linq.Expressions.NewExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.VisitBinary(BinaryExpression node)
at System.Linq.Expressions.BinaryExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
at System.Linq.Expressions.Compiler.VariableBinder.VisitInvocation(InvocationExpression node)
at System.Linq.Expressions.InvocationExpression.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.ExpressionVisitor.Visit(ReadOnlyCollection`1 nodes)
at System.Linq.Expressions.Compiler.VariableBinder.VisitLambda[T](Expression`1 node)
at System.Linq.Expressions.Expression`1.Accept(ExpressionVisitor visitor)
at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
at System.Linq.Expressions.Compiler.LambdaCompiler.Compile(LambdaExpression lambda, DebugInfoGenerator debugInfoGenerator)
at System.Linq.Expressions.Expression`1.Compile()
at LinqToDB.Linq.QueryRunner.Mapper`1.Map(IQueryRunner queryRunner, IDataReader dataReader)
at LinqToDB.Linq.QueryRunner.<ExecuteQuery>d__15`1.MoveNext()
```

Let's change the `q.Where(_ => _.SuperProduct == e)` to `q.Where(_ => _.SuperProductId == e.Id)`:
```C#
var q = Db.Context.GetTable<Product>();
Db.Context.GetTable<SuperProduct>()
	.Select(e => new
	{
		Fields = new
		{
			UiLabel = " " +
				e.NameEnUs,
			ProductCount = q.Where(_ => _.SuperProductId == e.Id).Count()
		},
	})
	.Where(_ => _.Fields.UiLabel.Contains("a")).Dump();
```

This will generate subqueries:
> There is already an open DataReader associated with this Command which must be closed first.
```
at System.Data.SqlClient.SqlInternalConnectionTds.ValidateConnectionForExecute(SqlCommand command)
at System.Data.SqlClient.SqlConnection.ValidateConnectionForExecute(String method, SqlCommand command)
at System.Data.SqlClient.SqlCommand.ValidateCommand(String method, Boolean async)
at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry)
at System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method)
at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method)
at System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior)
at System.Data.Common.DbCommand.System.Data.IDbCommand.ExecuteReader(CommandBehavior behavior)
at LinqToDB.Data.DataConnection.ExecuteReader(CommandBehavior commandBehavior)
at LinqToDB.Data.DataConnection.ExecuteReader()
at LinqToDB.Data.DataConnection.QueryRunner.ExecuteReader()
at LinqToDB.Linq.QueryRunner.ExecuteElement[T](Query query, IDataContextEx dataContext, Mapper`1 mapper, Expression expression, Object[] ps)
at LinqToDB.Linq.QueryRunner.<>c__DisplayClass25_0`1.<SetRunQuery>b__0(IDataContextEx db, Expression expr, Object[] ps)
at LinqToDB.Linq.ExpressionQuery`1.System.Linq.IQueryProvider.Execute[TResult](Expression expression)
at System.Linq.Queryable.Count[TSource](IQueryable`1 source)
at lambda_method(Closure , IQueryRunner , IDataReader )
at LinqToDB.Linq.QueryRunner.Mapper`1.Map(IQueryRunner queryRunner, IDataReader dataReader)
at LinqToDB.Linq.QueryRunner.<ExecuteQuery>d__15`1.MoveNext()
```

Let's remove `" " +`:
```C#
var q = Db.Context.GetTable<Product>();
Db.Context.GetTable<SuperProduct>()
	.Select(e => new
	{
		Fields = new
		{
			UiLabel = //" " +
				e.NameEnUs,
			ProductCount = q.Where(_ => _.SuperProductId == e.Id).Count()
		},
	})
	.Where(_ => _.Fields.UiLabel.Contains("a")).Dump();
```
This will work as expected.
Finally, let's revert `q.Where(_ => _.SuperProductId == e.Id)` back to `q.Where(_ => _.SuperProduct == e)`.
```C#
var q = Db.Context.GetTable<Product>();
Db.Context.GetTable<SuperProduct>()
	.Select(e => new
	{
		Fields = new
		{
			UiLabel = //" " +
				e.NameEnUs,
			ProductCount = q.Where(_ => _.SuperProduct == e).Count()
		},
	})
	.Where(_ => _.Fields.UiLabel.Contains("a")).Dump();
```
This also work properly.

UPDATED: Let's remove the ProductCount:
```C#
var q = Db.Context.GetTable<Product>();
Db.Context.GetTable<SuperProduct>()
	.Select(e => new
	{
		Fields = new
		{
			UiLabel = " " +
				e.NameEnUs,
			//ProductCount = q.Where(_ => _.SuperProduct == e).Count()
		},
	})
	.Where(_ => _.Fields.UiLabel.Contains("a")).Dump();
```
This also work.

### Environment details
linq2db version: *1.9.0*  
Database Server: *SQL Server 2016*  
Database Provider: *SqlServer2012*  
Operating system: *Windows 10* 
Framework version: *.NET Framework 4.7*

**Comment by n9** [comment]
Let's remove the wrapping Fields property:
```C#
var q = Db.Context.GetTable<Product>();
Db.Context.GetTable<SuperProduct>()
	.Select(e => new
	{
//		Fields = new
//		{
			UiLabel = " " +
				e.NameEnUs,
			ProductCount = q.Where(_ => _.SuperProduct == e).Count()
//		},
	})
	.Where(_ => _./*Fields.*/UiLabel.Contains("a")).Dump();
```
It also works properly.

**Comment by n9** [comment]
Removing the `Where(...)` also make it work:
```C#
var q = Db.Context.GetTable<Product>();
Db.Context.GetTable<SuperProduct>()
	.Select(e => new
	{
		Fields = new
		{
			UiLabel = " " +
				e.NameEnUs,
			ProductCount = q.Where(_ => _.SuperProduct == e).Count()
		},
	})
	//.Where(_ => _.Fields.UiLabel.Contains("a"))
	.Dump();
```

**Comment by n9** [comment]
@MaceWindu It looks like that is issue is no longer in any milestone.

**Comment by sdanyliv** [comment]
@n9, please use the following workarounds: https://github.com/linq2db/linq2db/issues/1198#issuecomment-398743915

**Comment by n9** [comment]
@sdanyliv You mean to use `Sql.Ext.Count` instead of `Count`?

**Comment by sdanyliv** [comment]
Oh, no. There is different query, let me think.

**Comment by MaceWindu** [comment]
@n9, sorry, was to impatient waiting for github response on moving issues to another milestone and messed it up a bit

**Comment by sdanyliv** [comment]
@n9, will handle this.

**Comment by n9** [comment]
Thx.

**Comment by sdanyliv** [comment]
@n9,  i have reviewed  your query i can't understand how it should work. 
So i have rewritten query in correct way. 
Note that i have tested on our tables and your sample transferred to the following query:
```cs
[Test, DataContextSource]
public void Test(string context)
{
	using (var db = GetDataContext(context))
	{
		var q = db.Child;

		var query = db.Parent
			.Select(e => new
			{
				Fields = new
				{
					Label = " " + e.Value1,
					Count = q.Where(_ => _.Parent == e).Count()
				},
			})
			.Where(_ => _.Fields.Label.Contains("1"));

		var array = query.ToArray();
	}
}
```
And there is correct query:
```cs
[Test, DataContextSource]
public void Test2(string context)
{
	using (var db = GetDataContext(context))
	{
		var q = db.Child;

		var childCount = from c in q
			group c by c.ParentID
			into g
			select new
			{
				ParentID = g.Key,
				Count = g.Count()
			};

		var query = (from e in db.Parent
				from c in childCount.InnerJoin(c => c.ParentID == e.ParentID)
				select new
				{
					Fields = new
					{
						Label = " " + e.Value1,
						Count = c.Count
					},
				})
			.Where(_ => _.Fields.Label.Contains("1"));

		var array = query.ToArray();
	}
}
```

**Comment by n9** [comment]
@sdanyliv, you suggestion looks a bit more complicated:) The strange think is that the following query works, although only `" " +` was removed.

```C#
var q = Db.Context.GetTable<Product>();
Db.Context.GetTable<SuperProduct>()
	.Select(e => new
	{
		Fields = new
		{
			UiLabel = //" " +
				e.NameEnUs,
			ProductCount = q.Where(_ => _.SuperProduct == e).Count()
		},
	})
	.Where(_ => _.Fields.UiLabel.Contains("a")).Dump();
``` 

What happened that such minor change breaks the linq2db?

**Comment by sdanyliv** [comment]
I know that it looks complicated but it how should be translated into SQL in correct way.
Show me the working SQL

**Comment by n9** [comment]
linq2sql generates the following SQL:

```SQL
-- Region Parameters
DECLARE @p0 NVarChar(1000) = ' '
DECLARE @p1 NVarChar(1000) = '%a%'
-- EndRegion
SELECT [t2].[value] AS [UiLabel], [t2].[value2] AS [ProductCount]
FROM (
    SELECT @p0 + [t0].[NameEnUs] AS [value], (
        SELECT COUNT(*)
        FROM [Products] AS [t1]
        WHERE [t1].[SuperProductId] = [t0].[Id]
        ) AS [value2]
    FROM [SuperProducts] AS [t0]
    ) AS [t2]
WHERE [t2].[value] LIKE @p1
```

**Comment by sdanyliv** [comment]
Interesting, which query has better execution plan. My or this one.

**Comment by sdanyliv** [comment]
BTW, i’ll check why we fail here. Thanks.

**Comment by n9** [comment]
Any chance to get this fixed? :) (Was removed from milestone again.)

**Comment by sdanyliv** [comment]
I had several try fix this. Not so easy, will find time for the next release. Sorry (

**Comment by n9** [comment]
@sdanyliv Thank you very much! (I was also trying to figure out what is going on, however I've realized that this part is really complicated for me to understand. It requires help from someone with deep linq2db knowledge.)

**Comment by n9** [comment]
@sdanyliv Have you found something there? Or could you please give me some hints how to fix that?

**Comment by sdanyliv** [comment]
I've tried, even found buggy line, but with no luck to fix it properly. Asking @igor-tkachev to support in such situation.

**Comment by sdanyliv** [comment]
@n9, finally i have found solution. Sorry for delay.

**Comment by n9** [comment]
Great, thanks:)

**Comment by MaceWindu** [comment]
Closed by #1399
